packages/core/src/combinator/merge.ts

```diff
@@ -34,29 +39,31 @@ export const mergeArray = streams =>
  * any nested Merge sources, in effect "flattening" nested
  * merge operations into a single merge.
  */
-const mergeStreams = streams =>
+const mergeStreams = <A>(streams: Stream<A>[]): Stream<A> =>
   streams.length === 0 ? empty()
     : streams.length === 1 ? streams[0]
-      : new Merge(reduce(appendSources, [], streams))
+      : new Merge(reduce<Stream<A>[], Stream<A>>(appendSources, [], streams))
 
-const withoutCanonicalEmpty = streams =>
+const withoutCanonicalEmpty = <A>(streams: ReadonlyArray<Stream<A>>): Stream<A>[] =>
   streams.filter(isNotCanonicalEmpty)
 
-const isNotCanonicalEmpty = stream =>
+const isNotCanonicalEmpty = <A>(stream: Stream<A>): boolean =>
   !isCanonicalEmpty(stream)
 
-const appendSources = (sources, stream) =>
+const appendSources = <A>(sources: Stream<A>[], stream: Stream<A>): Stream<A>[] =>
   sources.concat(stream instanceof Merge ? stream.sources : stream)
 
-class Merge {
-  constructor (sources) {
+class Merge<A> implements Stream<A> {
```

Comment:
It would be ideal if we could end up in a situation where `Merge`'s type reflects the (true) heterogeneous type of the merged streams (similar to what we've done with `Combine`).  If that proves to be tricky, then I'd be ok deferring it to another PR.

Reply:
I left a comment

Reply:
see https://github.com/mostjs/core/pull/296#discussion_r315147573